### PR TITLE
support source_path_from_root

### PR DESCRIPTION
**Pull request by AngryBerryMS**
[AB#341329](https://dev.azure.com/ceapex/Engineering/_workitems/edit/341329)

Support "source_path_from_root"

* No newly added warnings or errors.
* No probing through the repo.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6826)

### CODE REVIEW

File: docs/specs/redirection.yml [hunk 1]
```diff
@@ -764,3 +764,33 @@ outputs:
...
+            {"source_path": "b.md", "redirect_url": "/bb"}]
+        }
+      a/docfx.yml:
+      a/b/.openpublishing.redirection.json: |
+        {"redirections": [{"source_path_from_root": "/a/c.md", "redirect_url": "/aa"}]}
+outputs:
+  a/.publish.json: |
+    {
+      "files": [
+        {"url": "/a", "source_path": "a.md", "locale": "en-us", "redirect_url": "/aa"},
```
Comment (herohua) [inline docs/specs/redirection.yml:793]: where is the output for `b.md`?

File: src/docfx/build/redirection/RedirectionProvider.cs [hunk 2]
```diff
@@ -201,14 +206,38 @@ private static void GenerateRedirectionRules(ErrorBuilder errors, string docsetP
 
             foreach (var item in redirections.Concat(renames))
             {
-                if (item.SourcePath.IsDefault || string.IsNullOrEmpty(item.RedirectUrl))
+                if ((item.SourcePath.IsDefault && item.SourcePathFromRoot.IsDefault) || string.IsNullOrEmpty(item.RedirectUrl))
                 {
                     // Give a missing-attribute warning when source_path or redirect_url not specified
```
Comment (herohua) [inline src/docfx/build/redirection/RedirectionProvider.cs:211]: >source_path [](start = 61, length = 11)

update or just remove this comment

File: src/docfx/build/redirection/RedirectionProvider.cs [hunk 3]
```diff
@@ -201,14 +206,38 @@ private static void GenerateRedirectionRules(ErrorBuilder errors, string docsetP
...
+                }
+                else
+                {
+                    // TODO: syntax check
+                    var sourcePathRelativeToRepoRoot = item.SourcePathFromRoot.Value.Substring(1);
+                    if (repository != null)
+                    {
+                        sourcePath = Path.GetRelativePath(docsetPath, Path.Combine(repository.Path, sourcePathRelativeToRepoRoot));
+                    }
+                    else
```
Comment (herohua) [inline src/docfx/build/redirection/RedirectionProvider.cs:236]: >else [](start = 20, length = 4)

what is the case when repository == null?

**Review by herohua — CHANGES_REQUESTED** [review]
:clock1:

File: docs/specs/redirection.yml (hunk 1 above)
Comment (AngryBerryMS) [reply docs/specs/redirection.yml:793]: `b.md` is out of the build scope.

**Comment by xichan-ms** [comment]
> /azp run

For test usage, thanks.

File: src/docfx/build/redirection/RedirectionProvider.cs (hunk 3 above)
Comment (AngryBerryMS) [reply src/docfx/build/redirection/RedirectionProvider.cs:236]: For spec test.

File: src/docfx/build/redirection/RedirectionProvider.cs (hunk 2 above)
Comment (AngryBerryMS) [reply src/docfx/build/redirection/RedirectionProvider.cs:211]: Removed.

**Review by herohua — APPROVED** [review]

**Pull request merged**